micro-rdk-nmea/build.rs

```diff
@@ -0,0 +1,142 @@
+use serde_json::Value;
+use std::fs::{File, OpenOptions};
+use std::io::{BufReader, Write};
+use std::path::Path;
+
+use convert_case::{Case, Casing};
+
+// This is to compute a valid name for a variable or struct/enum member. We
+// want to remove special characters and introduce a default convention for
+// when a name starts with a number.
+fn clean_string_for_rust(input: &str, counter: usize) -> String {
+    let filtered: String = input
+        .chars()
+        .filter(|c| c.is_alphanumeric() || c.is_whitespace())
+        .collect();
+    if filtered
+        .chars()
+        .position(|c| c.is_numeric())
+        .map(|idx| idx == 0)
+        .unwrap_or_default()
+    {
+        // We want to use a prefix with an alphabetical suffix (i.e. "UnformattableVariantA")
+        // when the cleaned string still starts with a number. We map however many variants we've
+        // seen so far to an ASCII code in range 65-90 (A-Z). If there are more than 25 variants
+        // we want to use a second letter (i.e. "UnformattableVariantAB")
+        let suffix = if counter <= 25 {
+            ((counter + 65) as u8 as char).to_string()
+        } else {
+            // we assume that no enum has a number of variants > 25 * 25 = 625 (Note (GV) - I checked,
+            // this is true so far)
+            let first_letter_ascii = (((counter / 25) - 1) + 65) as u8;
+            let second_letter_ascii = ((counter % 25) + 65) as u8;
+            format!(
+                "{}{}",
+                first_letter_ascii as char, second_letter_ascii as char
+            )
+        };
+        format!("UnformattableVariant{}", suffix)
+    } else {
+        filtered
+    }
+}
+
+fn main() {
+    println!("cargo:rerun-if-changed=definitions.json");
+    println!("cargo::rustc-check-cfg=cfg(autogen_definitions)");
+
+    let enums_path = "src/gen/enums_gen.rs";
+
+    let mut enums_file = OpenOptions::new()
+        .create(true)
+        .truncate(true)
+        .write(true)
+        .read(true)
+        .open(enums_path)
+        .expect("could not create new enums_gen.rs");
+
+    enums_file
+        .write_all("// AUTO-GENERATED CODE; DO NOT DELETE OR EDIT\n".as_bytes())
+        .expect("failed to write warning statement");
+
+    let file_path = "definitions.json";
+    if !Path::new(file_path).exists() {
+        println!("No definitions file, skipping auto-generation...");
```

Comment:
what happen to the compile step when no definition are supplied? is the nmea driver even usable?

Reply:
The files are empty and the viamboat module doesn't even exist. So no it is not usable if compiled without definitions

Reply:
maybe we should not allow building if json is not present, if feature gating the nmea module isn't to complex

Reply:
the reason I didn't go in that direction is that I want to be able to run tests on the  individual data-type parsers